Support Securityadmin fixed server role

anju15bharti opened the pull request:
### Description

Added support for new fixed server role securityadmin

This role has following privileges with it:

1) Members of the securityadmin fixed server role can manage logins and their properties. 
2) They can GRANT, DENY, and REVOKE server-level permissions. 
3) securityadmin can also GRANT, DENY, and REVOKE database-level permissions if they have access to a database.

### Issues Resolved

[List any issues this PR will resolve]

### Test Scenarios Covered ###
* **Use case based -**


* **Boundary conditions -**


* **Arbitrary inputs -**


* **Negative test cases -**


* **Minor version upgrade tests -**


* **Major version upgrade tests -**


* **Performance tests -**


* **Tooling impact -**


* **Client tests -**



### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Support Securityadmin fixed server role

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

HarshLunagariya commented on contrib/babelfishpg_tds/src/backend/tds/tdsutils.c:
```diff
@@ -1229,18 +1231,18 @@ handle_grant_role(GrantRoleStmt *grant_stmt)
 			continue;
 
 		roleid = get_role_oid(rolename, false);
-		if (OidIsValid(roleid) && roleid == bbf_role_admin_oid)
```
This means we only want to allow `GRANT securityadmin TO ...` and `GRANT ... TO securityadmin` when we are in bbf_role_admin's context. Right?

anju15bharti replied:
Right or to be more specific, we restrict grant/revoke from pg side.

tanscorpio7 replied:
Do we also need to add these checks in is_babelfish_role() ?

anju15bharti replied:
yes

HarshLunagariya commented on contrib/babelfishpg_tsql/sql/ownership.sql:
```diff
@@ -285,11 +285,14 @@ BEGIN
 		RAISE E'Could not initialize babelfish with given role name: % is not the DB owner of current database.', sa_name;
 	END IF;
 
+	EXECUTE format('CREATE ROLE securityadmin CREATEROLE INHERIT PASSWORD NULL');
```
Do we need CREATEROLE privileges? Because internally we are doing it via the privileges of bbf_role_admin. Right now, I don't see a point of allowing CREATEROLE property because even if we make any login member of it, then also this property can't be inherited to it's members.

anju15bharti replied:
But a role can only grant this attribute if it itself has that attribute, hence we need to keep this attribute same as sysadmin.

tanscorpio7 replied:
But we are internally switching to bbf_role_admin which already has create role. No ?

anju15bharti replied:
Currently we are switching it to internal role but if in future we decide to use directly this, it should have the required privilege and also to explicitly reflect which priv it has and also to keep privilege wise similar to sysadmin and for future reference I have created this role with required attribute as discussed in design.

Deepesh125 replied:
@anju15bharti IMHO, let's not over engineer this feature and lets keep it simple for now. If such requirements comes up in future then we will revisit this. I am making this point especially in this feature it is sensitive feature in terms of security.

anju15bharti replied:
Agree, but we also don't need these attributes for sysadmin but we have kept it, so following that trend I have kept this attribute to this securityadmin role, Also giving this attribute to securityadmin is not escalating any privilege because even if we consider this role alone it should have this attribute.

HarshLunagariya commented on contrib/babelfishpg_tsql/sql/ownership.sql:
```diff
@@ -285,11 +285,14 @@ BEGIN
 	END IF;
 
+	EXECUTE format('CREATE ROLE securityadmin CREATEROLE INHERIT PASSWORD NULL');
+	EXECUTE format('GRANT CONNECT ON DATABASE %s TO securityadmin WITH GRANT OPTION', CURRENT_DATABASE());
```
1. Why granting connect permission is required? 
2. Acc. ideal T-SQL behaviour, Does all the members of securityadmin can login to database always -- even if some login is disabled? Otherwise it doesn't make sense to add this permission.

anju15bharti replied:
It's to grant connect permission on the babelfish physical database

tanscorpio7 replied:
Why do we need connect to babelfish physical database permission ?

anju15bharti replied:
We need this role to have atleast by default connect access on this babelfish physical database similar to our internal role, in fact for sysadmin, we provided create priv on this database which I think is necessarily not needed for our role we can just provide connect access to it.

HarshLunagariya replied:
Re-iterating again, But why does it actually need CONNECT? What are the internal operations are being blocked when we don't provide it?

anju15bharti replied:
It's only to have connect permission on the babelfish physical database or to be able to grant connect on physical database to other login but without explicitly granting connect to securityadmin it has connect permission. So need not to grant it.

HarshLunagariya commented on contrib/babelfishpg_tsql/sql/ownership.sql:
```diff
@@ -365,7 +371,8 @@ CAST(CASE WHEN Ext.type = 'R' THEN NULL ELSE Ext.credential_id END AS INT) AS cr
 CAST(CASE WHEN Ext.type = 'R' THEN 1 ELSE Ext.is_fixed_role END AS sys.BIT) AS is_fixed_role
 FROM pg_catalog.pg_roles AS Base INNER JOIN sys.babelfish_authid_login_ext AS Ext ON Base.rolname = Ext.rolname
-WHERE (pg_has_role(suser_id(), 'sysadmin'::TEXT, 'MEMBER')
+WHERE (bbf_is_member_of_role_nosuper(suser_id(), 'sysadmin') 
+  OR bbf_is_member_of_role_nosuper(suser_id(), 'securityadmin')
```
Should we use pg_has_role? Because anyway superuser will have all privileges/memberships.

tanscorpio7 replied:
+1

anju15bharti replied:
Right, but isn't it better to give the permission to the required role only?

tanscorpio7 replied:
I think we will get the same behaviour with `pg_has_role`.

anju15bharti replied:
will get same behaviour but the check will pass since it's superuser but here we don't want to check this, we want to check if login is member of sysadmin/securityadmin or not which will get not actually get tested since it will satisfy the superuser condition first.

HarshLunagariya commented on contrib/babelfishpg_tsql/sql/ownership.sql:
```diff
@@ -501,8 +508,20 @@ CAST(Ext.orig_loginname AS sys.nvarchar(128)) AS name,
+WHERE Ext.orig_loginname = 'sysadmin'
+AND Ext.type = 'R' AND
+bbf_is_member_of_role_nosuper(sys.suser_id(), 'sysadmin')
+UNION ALL
+SELECT
+CAST(Base.oid As INT) AS principal_id,
+CAST(CAST(Base.oid as INT) as sys.varbinary(85)) AS sid,
+CAST(Ext.orig_loginname AS sys.nvarchar(128)) AS name,
+CAST('SERVER ROLE' AS sys.nvarchar(128)) AS type,
+CAST ('GRANT OR DENY' as sys.nvarchar(128)) as usage
+FROM pg_catalog.pg_roles AS Base INNER JOIN sys.babelfish_authid_login_ext AS Ext ON Base.rolname = Ext.rolname
+WHERE Ext.orig_loginname = 'securityadmin'
+AND Ext.type = 'R' AND
+bbf_is_member_of_role_nosuper(sys.suser_id(), 'securityadmin');
```
Do we really need a separate union?

tanscorpio7 replied:
+1

anju15bharti replied:
Yes it can be optimized.

HarshLunagariya commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.3.0--4.4.0.sql:
```diff
@@ -11,6 +11,224 @@ SELECT set_config('search_path', 'sys, '||current_setting('search_path'), false)
+               SELECT FROM pg_catalog.pg_roles
+               WHERE  rolname = 'securityadmin') 
+        THEN
+            RAISE EXCEPTION 'Role "securityadmin" already exists.';
```
This exception will fail Latest-latest upgrade.

tanscorpio7 replied:
fixed roles are only targeted for PG 16, so should not be problem right ?

anju15bharti replied:
yes

HarshLunagariya commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y:
```diff
@@ -3145,11 +3158,24 @@ tsql_alter_server_role:
-			if (0 != strcmp($4, "sysadmin"))
+			
+			if (strcmp($4, "serveradmin") == 0
+				|| strcmp($4, "setupadmin") == 0
+				|| strcmp($4, "processadmin") == 0
+				|| strcmp($4, "dbcreator") == 0
+				|| strcmp($4, "diskadmin") == 0
+				|| strcmp($4, "bulkadmin") == 0) 
+			{
+					ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("Server role '%s' is currently not supported in Babelfish", $4),
+												parser_errposition(@4)));
+			}
+			else if (strcmp($4, "sysadmin") != 0 && strcmp($4, "securityadmin") != 0)
+			{
 				ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-                               errmsg("only sysadmin role is supported in ALTER SERVER ROLE statement"),
-                               parser_errposition(@4)));
+						errmsg("Only fixed server role is supported in ALTER SERVER ROLE statement"),
+									parser_errposition(@4)));
+			}
```
Let's avoid this duplicate code.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -2864,7 +2864,12 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
+						 * Check if the current login has privileges to create
+						 * login.
+						 */
+						if (!has_privs_of_role(GetSessionUserId(), get_role_oid("sysadmin", false)) &&
+																	!has_privs_of_role(GetSessionUserId(), get_securityadmin_oid()))
```
Please maintain 80 char limit for better readability.

HarshLunagariya commented on test/JDBC/expected/BABEL-2403.out:
```diff
@@ -81,6 +81,8 @@ name#!#pg_catalog#!#rolname#!#{"Rule": "<rolname> in babelfish_authid_login_ext
 text#!#sys#!#name#!#{"Rule": "<default_database_name> in babelfish_authid_login_ext must also exist in babelfish_sysdatabases"}
 name#!#pg_catalog#!#rolname#!#{"Rule": "<rolname> in babelfish_authid_login_ext must also exist in pg_authid"}
 text#!#sys#!#name#!#{"Rule": "<default_database_name> in babelfish_authid_login_ext must also exist in babelfish_sysdatabases"}
+name#!#pg_catalog#!#rolname#!#{"Rule": "<rolname> in babelfish_authid_login_ext must also exist in pg_authid"}
+text#!#sys#!#name#!#{"Rule": "<default_database_name> in babelfish_authid_login_ext must also exist in babelfish_sysdatabases"}
```
Why are these 2 entries appearing?

anju15bharti replied:
Because we have added one additonal login for securityadmin in the catalog.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -2430,3 +2455,25 @@ remove_createrole_from_logins(PG_FUNCTION_ARGS)
+		PG_RETURN_NULL();
+
+	member = PG_GETARG_OID(0);
+	rolename = text_to_cstring(PG_GETARG_TEXT_P(1));
+	dc_role = downcase_identifier(rolename, strlen(rolename), false, false);
+	role = get_role_oid(dc_role, false);
```
Can we pass the oid for rolename as well? To avoid down-casing and fetching the oid here.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -1628,8 +1649,12 @@ check_alter_server_stmt(GrantRoleStmt *stmt)
+	 * check if it has sysadmin privileges or
+	 * if server role is securityadmin and it has privileges of securityadmin
+	 */
+	if (!has_privs_of_role(GetSessionUserId(), sysadmin) && ((strcmp(granted_name, "securityadmin") != 0)
+										|| !has_privs_of_role(GetSessionUserId(), securityadmin_oid)))
```
What happens when we alter try to alter sysadmin role? For that case as well we should be throwing this kind of error. But from the current implementation, I don't think we will be throwing this error. Can you please validate and modify this check to include sysadmin check as well if required.

anju15bharti replied:
Yes it does include all the required check.

tanscorpio7 commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.3.0--4.4.0.sql:
```diff
@@ -11,6 +11,224 @@ SELECT set_config('search_path', 'sys, '||current_setting('search_path'), false)
  * final behaviour.
  */
 
+do
```
casing

tanscorpio7 commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y:
```diff
@@ -3127,12 +3127,25 @@ tsql_alter_server_role:
+				|| strcmp($4, "bulkadmin") == 0) 
+			{
+					ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("Server role '%s' is currently not supported in Babelfish", $4),
```
Server role --> Fixed server role ?

anju15bharti replied:
Since we only support fixed server, this means the same but will make it more specific if required.

tanscorpio7 commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y:
```diff
@@ -3127,12 +3127,25 @@ tsql_alter_server_role:
+							errmsg("Server role '%s' is currently not supported in Babelfish", $4),
+												parser_errposition(@4)));
+			}
+			else if (strcmp($4, "sysadmin") != 0 && strcmp($4, "securityadmin") != 0)
```
Can we move the check for supported server roles to epilogue file ?
This would make it reusable and we could also use MACROs instead of hard coded strings.

tanscorpio7 commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -615,6 +618,16 @@ get_bbf_role_admin_oid(void)
+get_securityadmin_oid(void)
+{
+	if (!OidIsValid(securityadmin_oid))
+		securityadmin_oid = get_role_oid("securityadmin", false);
```
Let us try to use MACRO everywhere

tanscorpio7 commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -1574,20 +1587,26 @@ bool
+	 * is alter server role statement, if the granted role is
 	 * server role
 	 */
 
+	bool is_sysadmin = 0;
+
 	if (list_length(stmt->granted_roles) == 1)
 	{
 		RoleSpec   *spec = (RoleSpec *) linitial(stmt->granted_roles);
+		int     	rolename_len = strlen(spec->rolename);
+
+		if (rolename_len == 8 && strncmp(spec->rolename, "sysadmin", 8) == 0)
+			is_sysadmin = true;
 
-		if (strcmp(spec->rolename, "sysadmin") == 0)	/* only supported server
-														 * role */
+		/* only supported server roles */
+		if (is_sysadmin || (rolename_len == 13 && strncmp(spec->rolename, "securityadmin", 13) == 0))
 			return true;
 	}
-	/* has one and only one grantee  */
-	if (list_length(stmt->grantee_roles) != 1)
+	/* if granted role is sysadmin and has one and only one grantee  */
+	if (is_sysadmin && list_length(stmt->grantee_roles) != 1)
 		return false;
```
We can remove these lines. We will anyway return false below.

anju15bharti replied:
Yes we can but since it's existing line, I didn't prefer to remove it.

tanscorpio7 replied:
I'd vote for removing it. Since we really want to maintain readability. Plus there would be lots of changes coming around this code soon.

HarshLunagariya replied:
Let's remove it.

tanscorpio7 commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -1574,20 +1587,26 @@ bool
-		if (strcmp(spec->rolename, "sysadmin") == 0)	/* only supported server
-														 * role */
+		/* only supported server roles */
+		if (is_sysadmin || (rolename_len == 13 && strncmp(spec->rolename, "securityadmin", 13) == 0))
```
Why are we using strncmp() ? I think we can simply use strcmp().

anju15bharti replied:
Keeping in mind that the alter role stmt also goes through this path, we should keep this check with strncmp instead strcmp since the rolename can be something different.

tanscorpio7 replied:
```
if (strcmp(spec->rolename, BABELFISH_SYSADMIN) == 0 || strcmp(spec->rolename, BABELFISH_SECURITYADMIN) == 0)
```
I meant using this check instead. Just to improve readability.

anju15bharti replied:
But strncmp is more preferrable over strcmp since strcmp is more vulnerable compared to strncmp.

tanscorpio7 commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -1604,9 +1623,11 @@ check_alter_server_stmt(GrantRoleStmt *stmt)
 
 	spec = (RoleSpec *) linitial(stmt->grantee_roles);
 	sysadmin = get_role_oid("sysadmin", false);
+	securityadmin_oid = get_securityadmin_oid();
```
We can directly call get_securityadmin_oid() when needed instead of keeping a local copy.

Commit:
Addressed Comments

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

anju15bharti committed:
Merge branch 'BABEL_4_X_DEV' into jira-babel-5040

Commit:
Updated expected_dependency file

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

tanscorpio7 commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -1591,20 +1604,26 @@ bool
+		int     	rolename_len = strlen(spec->rolename);
+
+		if (rolename_len == 8 && strncmp(spec->rolename, BABELFISH_SYSADMIN, 8) == 0)
+			is_sysadmin = true;
```
We can return from here itslelf, no need for using `is_sysadmin`.

anju15bharti replied:
Yes if we all agree to remove that existing line of code.

HarshLunagariya replied:
let's remove it.

tanscorpio7 commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -1645,8 +1664,12 @@ check_alter_server_stmt(GrantRoleStmt *stmt)
+	 * check if it has sysadmin privileges or
+	 * if server role is securityadmin and it has privileges of securityadmin
+	 */
+	if (!has_privs_of_role(GetSessionUserId(), sysadmin) && ((strcmp(granted_name, BABELFISH_SECURITYADMIN) != 0)
+										|| !has_privs_of_role(GetSessionUserId(), get_securityadmin_oid())))
```
2nd condition is a bit confusing, can we refactor it a bit ?

anju15bharti replied:
are we expecting to just give the negation of the second condition instead of expanding the negation?

tanscorpio7 replied:
Can `granted_name` be `sysadmin`. Should this be allowed for `BABELFISH_SECURITYADMIN ` members ?

anju15bharti replied:
This condition will not allow BABELFISH_SECURITYADMIN  members to alter sysadmin since it will pass the check and will throw error. I have testcases added as well.

tanscorpio7 commented on contrib/babelfishpg_tsql/sql/ownership.sql:
```diff
@@ -285,11 +285,14 @@ BEGIN
 	EXECUTE format('GRANT %I to bbf_role_admin WITH ADMIN TRUE;', sa_name);
 	EXECUTE format('CREATE ROLE sysadmin CREATEDB CREATEROLE INHERIT ROLE %I', sa_name);
 	EXECUTE format('GRANT sysadmin TO bbf_role_admin WITH ADMIN TRUE');
+	EXECUTE format('GRANT securityadmin TO bbf_role_admin WITH ADMIN TRUE');
```
We do not need a new statement, we can use ADMIN option to grant admin membership to bbf_role_admin when creating securityadmin.

anju15bharti replied:
I preferred creating the role clean and then granting required priv to required role as per the requirement.

Commit:
Added securityadmin membership check in is_babelfish_role

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
Merge remote-tracking branch 'upstream/BABEL_4_X_DEV' into jira-babel-5040

Commit:
rerun test

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
ignore logs for dump & restore

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Deepesh125 commented on contrib/babelfishpg_tds/src/backend/tds/tdsutils.c:
```diff
@@ -939,14 +939,17 @@ is_babelfish_role(const char *role)
 	Oid			bbf_msdb_guest_oid;
+	Oid			securityadmin_oid;
 
 	sysadmin_oid = get_role_oid(BABELFISH_SYSADMIN, true);	/* missing OK */
 	role_oid = get_role_oid(role, true);	/* missing OK */
+	securityadmin_oid = get_role_oid(BABELFISH_SECURITYADMIN, true);  /* missing OK */
```
Lets use `get_sysadmin_oid` to get oid of sysadmin and also introduce similar function to get oid of security_admin

anju15bharti replied:
we don't have get_sysadmin_oid in tds, we have get_securityadmin_oid and get_sysadmin_oid in tsql

tanscorpio7 replied:
We should revisit this function `is_babelfish_role`.
I am curious why we cannot simply check if bbf_admin_role is member of the role to verify if it is a babelfish role.

Deepesh125 commented on contrib/babelfishpg_tsql/sql/ownership.sql:
```diff
@@ -501,8 +508,11 @@ CAST(Ext.orig_loginname AS sys.nvarchar(128)) AS name,
 FROM pg_catalog.pg_roles AS Base INNER JOIN sys.babelfish_authid_login_ext AS Ext ON Base.rolname = Ext.rolname
-WHERE Ext.type = 'R' AND
-(pg_has_role(sys.suser_id(), 'sysadmin'::TEXT, 'MEMBER'));
+WHERE Ext.type = 'R'
+AND Ext.orig_loginname = 'sysadmin'
+AND bbf_is_member_of_role_nosuper(sys.suser_id(), sys.suser_id('sysadmin'))
+OR Ext.orig_loginname = 'securityadmin'
+AND bbf_is_member_of_role_nosuper(sys.suser_id(), sys.suser_id('securityadmin'));
```
Can you explain this change for my understanding?

anju15bharti replied:
login_token should return rows for fixed server role of which the current login is member of.

HarshLunagariya replied:
Then why do we need `Ext.orig_loginname = 'xyz'` condition. Don't we just need to check whether it is a fixed server role i.e. `type = 'R' AND is_fixed_role =1`?

tanscorpio7 replied:
A superuser will anyway have select privileges on underlying pg_catalogs. I do not think we need to restrict super users here.

anju15bharti replied:
@tanscorpio7 We intentionally added the API which should not consider superuser priv because if superuser is also considers then in BBF for master user if it is only member of sysadmin then also it will bypass the check for securityadmin membership as well and we'll see row for securityadmin as well which is not expected in sql server.I will confirm with dbes.

Deepesh125 commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -2472,3 +2495,21 @@ remove_createrole_from_logins(PG_FUNCTION_ARGS)
+
+PG_FUNCTION_INFO_V1(bbf_is_member_of_role_nosuper);
+Datum
+bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
```
is there any alternate way of doing this? defining C function just to check membership may result in performance penalty. @sumitj824 can give more insights on this

tanscorpio7 replied:
We do not need to restrict superusers from viewing `sys` catalogs. They anyway have access to everything in pg_catalog.

Deepesh125 commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -1967,3 +1967,24 @@ tsql_index_nulls_order(List *indexParams, const char *accessMethod)
 }
+
+static void
+is_server_role_supported (const char* serverrole, int position, core_yyscan_t yyscanner)
```
nitpick: const char * serverrole

anju15bharti replied:
changed

Deepesh125 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3140,7 +3147,8 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
 							if (strcmp(defel->defname, "password") == 0)
 							{
-								if (get_role_oid(stmt->role->rolename, true) != GetSessionUserId() && !is_member_of_role(GetSessionUserId(), datdba))
+								if (get_role_oid(stmt->role->rolename, true) != GetSessionUserId() && (!is_member_of_role(GetSessionUserId(), datdba)
+																					&& !is_member_of_role(GetSessionUserId(), securityadm_oid)))
 									ereport(ERROR,(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
 											 errmsg("Cannot alter the login '%s', because it does not exist or you do not have permission.", stmt->role->rolename)));
```
doesnt this seem like redundant?

tanscorpio7 replied:
+1, we already check for member ship just before

anju15bharti replied:
It is not actually redundant it's just that password based login case is handled separately and before the general case, below that we are not checking if it's password based login.

Deepesh125 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3705,13 +3727,30 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
+					}
+
+					/* Otherwise, provide attribute for role priv */
 					else
```
just completeness, check `else if rolspec->rolename == "security_admin" then .... else throw error`

anju15bharti replied:
we have blocked alter server role for any other roles other than sysadmin and securityadmin in the parsing phase itself so this else condition which you are taking about will actually never hit.

HarshLunagariya replied:
@anju15bharti Agree. But when you support dbcreator role in your next PR then anyway you're modify this check , right? Let's do it here only.

anju15bharti replied:
will modify this check in dbcreator but there also will not add any else condition because that according to me is irrelevant since it will never be the case. On next PR we will just add additional else check to handle another case.

tanscorpio7 replied:
@anju15bharti ^

Deepesh125 commented on contrib/babelfishpg_tsql/src/procedures.c:
```diff
@@ -3018,6 +3020,16 @@ sp_addlinkedsrvlogin_internal(PG_FUNCTION_ARGS)
 	initStringInfo(&query);
 
+	/*
+	 * check privileses for login
```
please fix spelling mistake

anju15bharti replied:
corrected

Deepesh125 commented on contrib/babelfishpg_tsql/src/procedures.c:
```diff
@@ -3018,6 +3020,16 @@ sp_addlinkedsrvlogin_internal(PG_FUNCTION_ARGS)
+				!has_privs_of_role(GetSessionUserId(), get_securityadmin_oid()))
+		ereport(ERROR,
+				(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+					errmsg("User does not have permission to perform this action.")));
```
Please point out additional testing being done for this change

anju15bharti replied:
test/JDBC/expected/securityadmin_role-vu-verify.out:222

Deepesh125 commented on test/JDBC/expected/is_srvrolemember.out:
```diff
@@ -122,7 +122,7 @@ SELECT is_srvrolemember('securityadmin')
 ~~START~~
 int
-0
+1
```
where are we granting this permission?

Deepesh125 commented on test/JDBC/expected/is_srvrolemember-vu-verify.out:
```diff
@@ -122,7 +122,7 @@ SELECT is_srvrolemember('securityadmin')
 ~~START~~
 int
-0
+1
```
where are we granting this permission?

anju15bharti replied:
contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql:2029

Deepesh125 commented on test/JDBC/expected/securityadmin_role-vu-verify.out:
```diff
@@ -0,0 +1,1268 @@
+-- tsql
```
We should add test cases for following scenarios - 
1. Test cases in conjunction with GRANT ON SCHEMA support 
2. add test cases where user is member of both `sysadmin` and `securityadmin`, it should be able to connect to newly created database
3. we should also add negative tests cases where we try to add built in users to `securityadmin`
4. `create login securityadmin with password = '123'` should be error'd out
5. `alter server role securityadmin add member <db-role>` -- this should be error'd out also
6.

anju15bharti replied:
Added.

Commit:
Added more testcases

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

anju15bharti committed:
Merge branch 'BABEL_4_X_DEV' into jira-babel-5040

Commit:
Fix test failure

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
Revert "ignore logs for dump & restore"

This reverts commit c584f245804f1b7017eec64ca20cf2da619a2e4d.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -1967,3 +1967,24 @@ tsql_index_nulls_order(List *indexParams, const char *accessMethod)
 }
+
+static void
+is_server_role_supported (const char *serverrole, int position, core_yyscan_t yyscanner)
```
better naming required. It should indicate that it is throwing error.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3715,14 +3762,16 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
 							standard_ProcessUtility(pstmt, queryString, readOnlyTree, context, params,
 													queryEnv, dest, qc);
-						exec_alter_role_cmd(query.data, spec);
+						if (*query.data)
```
Why `*query.data`?

anju15bharti replied:
because the *query.data can be null if it's member of sysadmin then we do not need to execute the internal revoking command.

tanscorpio7 replied:
We should be checking for length here `query.len`

shalinilohia50 replied:
+1. We should have a better check with explanation.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3692,13 +3714,38 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
-						appendStringInfo(&query, "ALTER ROLE dummy WITH createrole createdb; ");
+					
+					/* If sysadmin, provide attribute for role and database priv */
+					if (strlen(rolspec->rolename) == 8 && strncmp(rolspec->rolename, BABELFISH_SYSADMIN, 8) == 0)
```
Should we use some macro/function for `strlen(rolspec->rolename) == 8 && strncmp(rolspec->rolename, BABELFISH_SYSADMIN, 8) == 0` it is at multiple places.

HarshLunagariya commented on test/JDBC/expected/securityadmin_role-vu-verify.out:
```diff
@@ -0,0 +1,1520 @@
+varchar#!#varchar
+sysadmin#!#jdbc_user
+securityadmin#!#securityadmin_login1
+~~END~~
```
Does it shows all the logins? If yes, then let's show only the logins which we want to validate. Because stale logins from other tests can affect it's output in future.

HarshLunagariya commented on test/JDBC/expected/securityadmin_role-vu-verify.out:
```diff
@@ -0,0 +1,1520 @@
+~~ERROR (Message: 'GRANT Database' is not currently supported in Babelfish)~~
+
+
+GRANT ALTER ANY EVENT NOTIFICATION TO no_securityadmin_login1 WITH GRANT OPTION; 
+go
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: 'GRANT Database' is not currently supported in Babelfish)~~
```
This message this misleading, But it can be fixed separately if required in separate effort.

HarshLunagariya commented on test/JDBC/expected/securityadmin_role-vu-verify.out:
```diff
@@ -0,0 +1,1520 @@
+
+
+-- System objects
+-- All rows of server_prinicipals view should be visible to securityadmin login
+-- limited rows should get displayed
+SELECT name, type, type_desc, default_database_name, default_language_name, credential_id, owning_principal_id, is_fixed_role
+FROM sys.server_principals 
+WHERE name =  'jdbc_user';
+GO
+~~START~~
+varchar#!#char#!#nvarchar#!#varchar#!#varchar#!#int#!#int#!#bit
+jdbc_user#!#S#!#SQL_LOGIN#!#master#!#English#!#-1#!#-1#!#0
+~~END~~
+
+
+SELECT name, type, type_desc, default_database_name, default_language_name, credential_id, owning_principal_id, is_fixed_role
+FROM sys.server_principals 
+WHERE name =  'sysadmin';
+GO
+~~START~~
+varchar#!#char#!#nvarchar#!#varchar#!#varchar#!#int#!#int#!#bit
+sysadmin#!#R#!#SERVER_ROLE#!#<NULL>#!#English#!#<NULL>#!#1#!#1
+~~END~~
+
+
+SELECT name, type, type_desc, default_database_name, default_language_name, credential_id, owning_principal_id, is_fixed_role
+FROM sys.server_principals 
+WHERE name =  'public';
+GO
+~~START~~
+varchar#!#char#!#nvarchar#!#varchar#!#varchar#!#int#!#int#!#bit
+public#!#R#!#SERVER_ROLE#!#<NULL>#!#<NULL>#!#<NULL>#!#1#!#0
+~~END~~
+
```
nit: There are multiple such queries starting from 'jdbc_user' to '%securityadmin%'. Let's just combine them to save some test execution time.

HarshLunagariya commented on test/JDBC/expected/securityadmin_role-vu-verify.out:
```diff
@@ -0,0 +1,1520 @@
+-- terminate-tsql-conn user=securityadmin_login1 password=123
+
+-- psql
+SET client_min_messages = 'error';
```
Why this?

HarshLunagariya commented on test/JDBC/expected/securityadmin_role-vu-verify.out:
```diff
@@ -0,0 +1,1520 @@
+EXEC sp_droplinkedsrvlogin 'Accounts', NULL
+GO
+
+-- inside procedure
```
Let's update the test plan to include these cases as well.

HarshLunagariya commented on test/JDBC/expected/securityadmin_role-vu-verify.out:
```diff
@@ -0,0 +1,1520 @@
+
+-- terminate-tsql-conn user=securityadmin_login1 password=123
+
+-- tsql
+-- login is member of both securityadmin as well as sysadmin
```
Not mentioned in test plan, let's include this family of test cases as well.

HarshLunagariya commented on test/JDBC/expected/securityadmin_role-vu-verify.out:
```diff
@@ -0,0 +1,1520 @@
+-- tsql user=securityadmin_login1 password=123
+
+-- both attribute should be true
+select rolname, rolcreaterole, rolcreatedb from pg_roles where rolname = 'securityadmin_login1'
+go
+~~START~~
+varchar#!#bit#!#bit
+securityadmin_login1#!#1#!#1
+~~END~~
```
can we test the `is_srvrolemember` for this case as well where it is member of sysadmin but not securityadmin? If we haven't add already at anywhere else.

anju15bharti replied:
It's added below.

HarshLunagariya commented on test/JDBC/expected/securityadmin_role-vu-verify.out:
```diff
@@ -0,0 +1,1520 @@
+~~END~~
+
+
+select bbf_is_member_of_role_nosuper(suser_id(), suser_id('securityadmin'))
+go
+~~START~~
+bit
+0
+~~END~~
+
+
+select bbf_is_member_of_role_nosuper(suser_id(), suser_id('sysadmin'))
+go
+~~START~~
+bit
+1
+~~END~~
```
Execute `is_srvrolemember()` as well.

HarshLunagariya commented on test/JDBC/expected/securityadmin_role-vu-verify.out:
```diff
@@ -0,0 +1,1520 @@
+    Server SQLState: 42501)~~
+
+
+GRANT securityadmin TO securityadmin_restrict_new_login;
```
Try granting it to PG created role.

HarshLunagariya commented on test/JDBC/expected/securityadmin_role-vu-verify.out:
```diff
@@ -0,0 +1,1520 @@
+    Server SQLState: 42501)~~
+
+
+SET SESSION AUTHORIZATION securityadmin;
+GO
+~~ERROR (Code: 0)~~
+
+~~ERROR (Message: ERROR: permission denied to set session authorization
+    Server SQLState: 42501)~~
+
+
+SET ROLE securityadmin;
+GO
+~~ERROR (Code: 0)~~
+
+~~ERROR (Message: ERROR: permission denied to set role "securityadmin"
+    Server SQLState: 42501)~~
```
Which all roles can set it? Should we block it explicitly (w/o consulting permissions)?

anju15bharti replied:
From pg side, no babelfish role should be allowed to set if user is not superuser.

HarshLunagariya commented on test/JDBC/expected/securityadmin_role-vu-verify.out:
```diff
@@ -0,0 +1,1520 @@
+
+~~ERROR (Message: ERROR: permission denied to set role "securityadmin"
+    Server SQLState: 42501)~~
+
```
Add testcases where you are trying to assign object ownership to securityadmin.

Commit:
Addressed comments

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
Merge remote-tracking branch 'upstream/server-database-roles-4-x' into jira-babel-5040

anju15bharti committed:
Merge branch 'server-database-roles-4-x' into jira-babel-5040

Commit:
Retrigger Tests

Commit:
Fixed test failure

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

tanscorpio7 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3505,8 +3526,9 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
 						if (!OidIsValid(role_oid) ||
-							!is_member_of_role(GetSessionUserId(), get_sysadmin_oid()) ||
-							role_oid == get_bbf_role_admin_oid())
+							(!is_member_of_role(GetSessionUserId(), get_sysadmin_oid()) && 
+							!is_member_of_role(GetSessionUserId(), securityadmin_oid)) ||
```
We already check for membership before at line 3507

tanscorpio7 replied:
Instead let us add check for sysadmin and restrict its drop here. Currently we rely on postgres on restricting dropping sysadmin.
I could run this on psql endpoint
```
drop owned by sysadmin;
```
and then drop sysadmin from TDS endpoint.

Deepesh125 replied:
good catch! we need some common logic to prevent alteration of pre-defined roles

anju15bharti replied:
we have already blocked for securityadmin and bbf_role_admin role to get dropped, will add the same for sysadmin as well.

tanscorpio7 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3723,14 +3770,16 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
 						SetUserIdAndSecContext(save_userid, save_sec_context);
-						pfree(query.data);
+						if (query.data)
+							pfree(query.data);
```
This will not free the memory if query.data is empty string. But we should unconditionally free it, since `initStringInfo` always allocates memory.

tanscorpio7 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -2912,7 +2912,12 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
+						/*
+						 * Check if the current login has privileges to create
+						 * login.
+						 */
+						if (!has_privs_of_role(GetSessionUserId(), get_sysadmin_oid()) &&
+								!has_privs_of_role(GetSessionUserId(), get_securityadmin_oid()))
```
Alter login should not work if current user is member of security admin and the login being altered is member of sysadmin

HarshLunagariya replied:
Please validate and add test for the same as well.

tanscorpio7 commented on test/JDBC/input/securityadmin_role-vu-prepare.mix:
```diff
@@ -0,0 +1,90 @@
+create role securityadmin_rol
+go
+
+create database securityadmin_db1
+go
```
Lets also add a test case where securityadmin tries to drop a database it does not own and fails.

anju15bharti replied:
we have testcase added in test/JDBC/expected/securityadmin_role-vu-verify.out:625

shalinilohia50 commented on test/JDBC/expected/securityadmin_role-vu-verify.out:
```diff
@@ -0,0 +1,1557 @@
+go
+
+-- make login member of securityadmin
+Alter server role securityadmin add member securityadmin_login1
```
If a login already has a user/no user, can it be added as a member of `securityadmin`? 
Let's add a test for the same.

anju15bharti replied:
In sysadmin, we additionally added check to not allow user creation or block making member if has user mapped but here we haven't even added these additional check then why is it needed in the file? Although I have already tested it locally.

Deepesh125 commented on contrib/babelfishpg_tsql/src/procedures.c:
```diff
@@ -3103,33 +3131,59 @@ sp_droplinkedsrvlogin_internal(PG_FUNCTION_ARGS)
+		* implementation of linked logins, we try to drop USER MAPPINGs for both the CURRENT_USER
+		* and PUBLIC PG roles.
+		*/
+		appendStringInfo(&query, "DROP USER MAPPING IF EXISTS FOR CURRENT_USER SERVER \"%s\"", servername);
```
make sure to free this

pranavJ23 committed:
fixing the issue of select cast with invalid number of function argument (#3005)

Cache lookup was failing when we ran cast query with extensions like PgAudit enabled.
The issue was because the arguments were not passed correctly into the cast functions. Added the handling cases as per the number of arguments(nargs) passed via cast query.

Issues Resolved : [BABEL-5272] [BABEL-4476]


Signed-off-by: Pranav Jain pranavjc@amazon.com

Yvinayak07 committed:
Fixed UDT handling for CASE statement when branch expression is of String Datatype. (#3009)

In CASE expression if we use UDT or sysname in any of the branch, there was no handling to find overall common type
and common typmod for CASE statement. For example, 

create type BABEL_CASE_EXPR_TEST_UDT from NVARCHAR(100);
GO

SELECT 
CASE 2
   WHEN 1 THEN 'abc'
   WHEN 2 THEN CAST('char' AS BABEL_CASE_EXPR_TEST_UDT)
END AS RESULT
GO

output:

Msg 8152, Level 16, State 1, Server BABELFISH, Line 1
value too long for type character varying(3)

Expected output:

nvarchar
char

Changes made to fix the issues - 

We added handling for UDT and sysname for CASE Expression. While finding common type of CASE Expression, we
checked if the following branch is a UDT or not, if yes then find the immediate base_type of UDT and find the TSQL
precedence according to the immediate base_type. If the branch if of type SYSNAME then we directly assign varchar as
datatype for that branch. While finding the common_typmod of CASE Expression, We checked if the branch is a UDT or
not, if yes then find the typmod of the base_type using getBaseTypeAndTypmod().If one of the branch is
of type SYSNAME then set typmod as SYSNAME_TYPMOD (i.e. 128).

Task: BABEL-5103, BABEL-4332
Signed-off-by: yashneet vinayak <yashneet@amazon.com>

ahmed-shameem committed:
Add '\' as ESCAPE character for sp_columns_100 for table_owner parameter (#3014)

There are few occurrences where we use procedure sp_columns_100 to fetch column metadata. The issue with this procedure is, when the table_owner (schema name) contains _ character, we may obtain some unexpected output . This is because in some places we are using using sp_columns_100 incorrectly with a backslash escaping underscores (e.g. “_“). Underscores are wildcards in sp_columns_100 (and in LIKE in general), but sp_columns_100 does not use a backslash as an escape character.
Till now, sp_columns_100 , was invoking a plpgsql function sp_column_100_internal which used PG’s LIKE operator for table_owner, which does use a backslash as the default escape character.

With BABEL-4953, we have updated the definition of sp_columns_100 where we removed redundant function call and implemented the procedure by using TSQL LIKE operator on column name, which will create some problems in other places.

The change in the new version is correct. However, since currently we rely on the incorrect functionality of the escape characters, we are making this temporary change for now. Once, we have fixed the dependencies on this procedure, we need to revert it.

Authored-by: Shameem Ahmed shmeeh@amazon.com
Signed-off-by: Shameem Ahmed shmeeh@amazon.com

Yvinayak07 committed:
Updating the upgrade test file after fixing the typmod for CASE expression (#3015)

Updating the upgrade test file after fixing the typmod for CASE expression. Need to change the 15_9 schedule file in
BABEL_4_X_DEV, as PR for BABEL_4_X_DEV is merged first, Which includes before files in 15_9 schedule file. But as
these changes need to be merged in BABEL_3_X_DEV, we need to add after files in place of before files as 15_9
corresponds to BABEL_3_X_DEV.

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Commit:
Added testfile to upgrade script & handled alter login(sysadmin member) case

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

tanscorpio7 committed:
fix valgrind warning in pltsql_store_view_definition (#3019)

Reset static variable original_query_string back to NULL after use otherwise next queries which do not set original_query_string in ANLTR will read already freed memory.
---------
Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

robverschoor committed:
Support unary '+' operators for strings (#3013)


The T-SQL grammar allows an arbitrary number of unary + operators to precede an expression,
but PG only supports that for numeric expressions. For string expressions, the + will be parsed fine by the T-SQL ANTLR parser, but these will raise an error in PG.
In SQL those unary + string operators may look like they are additional concatenation operators, like: SELECT 'a' ++ 'b'. Expressions such as +++(+++@v)) are also valid syntax according to the T-SQL grammar even though they look unusual.
In this fix we remove such unary + operators, which are redundant anyway.
However we do not touch numeric constants (e.g. +123) since the +, though still redundant, may
have been included for code clarity (e.g. +123 as opposed to -123).

Signed-off-by: Rob Verschoor rcv@amazon.com

Issues Resolved
BABEL-1924 Support unary '+' operator for string expressions

Jakeowen1 committed:
Add support for Alter Function on multi-line table value functions (#2999)


This commit adds the grammar rules and logic to support alter function for multi-statement table value functions cases. The additional grammar rule is included with the other tsql_AlterFunctionStmt cases, and follows the syntax for creating multi line tvfs. The logic is the same as the other alter function cases, with the addition of handling for tabletypstmt, which represents the underlying relationship of these functions.

Issues Resolved: BABEL-5247
Signed-off-by: Jake Owen <owjco@amazon.com>

Deepesh125 commented on test/JDBC/expected/securityadmin_role-vu-verify.out:
```diff
@@ -0,0 +1,1609 @@
+
+
+-- alter server role securityadmin add member db roles should error out
+Alter server role securityadmin add member guest
+go
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: role "guest" does not exist)~~
```
This error is misleading

anju15bharti replied:
since alter server role is server level and on server level guest is not present, so it is expected.

Commit:
Merge remote-tracking branch 'origin/server-database-roles-4-x' into jira-babel-5040

anju15bharti commented:
Closing this PR since the target branch is moved to amazon aurora repo
New PR https://github.com/amazon-aurora/babelfish_extensions/pull/67
